File: src/main/java/org/taktik/icure/dao/impl/GenericDAOImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.dao.impl;

import org.apache.commons.lang3.ArrayUtils;
import org.ektorp.AttachmentInputStream;
import org.ektorp.BulkDeleteDocument;
import org.ektorp.CouchDbConnector;
import org.ektorp.DocumentNotFoundException;
import org.ektorp.DocumentOperationResult;
import org.ektorp.Options;
import org.ektorp.UpdateConflictException;
import org.ektorp.ViewQuery;
import org.ektorp.support.DesignDocument;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Qualifier;
import org.taktik.icure.dao.GenericDAO;
import org.taktik.icure.dao.Option;
import org.taktik.icure.dao.impl.ektorp.CouchDbICureConnector;
import org.taktik.icure.dao.impl.idgenerators.IDGenerator;
import org.taktik.icure.dao.impl.idgenerators.UUIDGenerator;
import org.taktik.icure.dao.impl.keymanagers.KeyManager;
import org.taktik.icure.dao.impl.keymanagers.UniversallyUniquelyIdentifiableKeyManager;
import org.taktik.icure.entities.base.StoredDocument;
import org.taktik.icure.exceptions.BulkUpdateConflictException;

import javax.persistence.PersistenceException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

public abstract class GenericDAOImpl<T extends StoredDocument> extends CouchDbICureRepositorySupport<T> implements GenericDAO<T> {
	private static final Logger log = LoggerFactory.getLogger(GenericDAOImpl.class);

	protected final Class<T> entityClass;
	protected final KeyManager<T, String> keyManager;
	protected UUIDGenerator uuidGenerator;

	public void setUuidGenerator(UUIDGenerator uuidGenerator) {
		this.uuidGenerator = uuidGenerator;
	}

	public GenericDAOImpl(Class<T> entityClass, @Qualifier("couchdb") CouchDbICureConnector db, IDGenerator idGenerator) {
		super(entityClass, db);
		this.entityClass = entityClass;
		this.keyManager = new UniversallyUniquelyIdentifiableKeyManager<>(idGenerator);
	}

	protected Map<String, Object> getStorageOptions() {
		return null;
	}

	@Override
	public boolean contains(String id) {
		if (log.isDebugEnabled()) {
			log.debug(entityClass.getSimpleName() + ".contains: " + id);
		}
		return super.contains(id);
	}

	@Override
	public boolean hasAny() {
		return designDocContainsAllView() && db.queryView(createQuery("all").limit(1)).getSize() > 0;
	}

	@Override
	public List<T> getAll() {
		if (log.isDebugEnabled()) {
			log.debug(entityClass.getSimpleName() + ".getAll");
		}
		List<T> result = super.getAll();

		result.forEach(this::postLoad);

		return result;
	}

	@Override
	public String getAttachment(String documentId, String attachmentId) {
		AttachmentInputStream attachmentInputStream = db.getAttachment(documentId, attachmentId);
		BufferedReader reader = new BufferedReader(new InputStreamReader(attachmentInputStream));

		@SuppressWarnings("MismatchedQueryAndUpdateOfStringBuilder") StringBuilder sb = new StringBuilder();
		reader.lines().forEach(sb::append);
		return sb.toString();
	}

	@Override
	public AttachmentInputStream getAttachmentInputStream(String documentId, String attachmentId, String rev) {
		return rev != null ? db.getAttachment(documentId, attachmentId, rev):db.getAttachment(documentId, attachmentId);
	}

	@Override
	public String createAttachment(String documentId, String rev, AttachmentInputStream data) {
		// return Document Revision
		return db.createAttachment(documentId, rev, data);
	}

	@Override
	public String deleteAttachment(String documentId, String rev, String attachmentId) {
		// return Document Revision
		return db.deleteAttachment(documentId, rev, attachmentId);
	}

	@Override
	public T get(String id, Option... options) {
		if (log.isDebugEnabled()) {
			log.debug(entityClass.getSimpleName() + ".get: " + id + " [" + ArrayUtils.toString(options) + "]");
		}
		try {
			T result = super.get(id, asEktorpOptions(options));

			postLoad(result);

			return result;
		} catch (DocumentNotFoundException e) {
			log.warn("Document not found",e);
		}
		return null;
	}

	public T get(String id, String rev) {
		try {
			T result = super.get(id, rev);

			postLoad(result);

			return result;
		} catch (DocumentNotFoundException e) {
			log.warn("Document not found",e);
		}
		return null;
	}

	public T get(String id) {
		try {
			T result = super.get(id);

			postLoad(result);

			return result;
		} catch (DocumentNotFoundException e) {
			log.warn("Document not found",e);
		}
		return null;
	}


	@Override
	public Set<T> getSet(Collection<String> ids) {
		if (log.isDebugEnabled()) {
			log.debug(entityClass.getSimpleName() + ".get: " + ids);
		}
		return new HashSet<>(getList(ids));
	}

	@Override
	public List<T> getList(Collection<String> ids) {
		if (log.isDebugEnabled()) {
			log.debug(entityClass.getSimpleName() + ".get: " + ids);
		}
		ViewQuery q = new ViewQuery()
				.allDocs()
				.includeDocs(true)
				.keys(ids);
		q.setIgnoreNotFound(true);
		List<T> result = queryResults(q);

		result.forEach(this::postLoad);

		return result;
	}

	@Override
	public T newInstance() {
		// Instantiate new entity
		T entity;
		try {
			entity = entityClass.newInstance();
		} catch (InstantiationException | IllegalAccessException e) {
			throw new RuntimeException("Could not instantiate entity of class " + entityClass.getName(), e);
		}

		// Set new key
		keyManager.setNewKey(entity, entityClass.getSimpleName());

		return entity;
	}

	@Override
	public T create(T entity) {
		return save(true, entity);
	}

	@Override
	public T save(T entity) {
		return save(null, entity);
	}

	protected T save(Boolean newEntity, T entity) {
		if (entity != null) {
			if (log.isDebugEnabled()) {
				log.debug(entityClass.getSimpleName() + ".save: " + entity.getId() + ":" + entity.getRev());
			}

			// Before save
			beforeSave(entity);

			// Check if key is missing and if this is a new entity
			boolean missingKey = (entity.getId() == null);

			// Add new key if missing
			if (missingKey) {
				keyManager.setNewKey(entity, entityClass.getSimpleName());
				newEntity = true;
			} else {
				if (newEntity == null) {
					newEntity = entity.getRev() == null;
				}
			}

			if (!newEntity) {
				//saveRevHistory(entity, null);
			}

			// Save entity
			super.update(entity);

			// After save
			afterSave(entity);
		}

		return entity;
	}

	protected void beforeSave(T entity) {
	}

	protected void afterSave(T entity) {
	}

	@Override
	public void remove(T entity) {
		if (entity != null) {
			if (log.isDebugEnabled()) {
				log.debug(entityClass.getSimpleName() + ".remove: " + entity);
			}
			// Before remove
			beforeDelete(entity);
			// Delete
			super.remove(entity);
			// After remove
			afterDelete(entity);
		}
	}

	@Override
	public void unremove(T entity) {
		if (entity != null) {
			if (log.isDebugEnabled()) {
				log.debug(entityClass.getSimpleName() + ".unremove: " + entity);
			}
			beforeUnDelete(entity);
			super.unremove(entity);
			afterUnDelete(entity);
		}
	}

	@Override
	public void purge(T entity) {
		if (entity != null) {
			if (log.isDebugEnabled()) {
				log.debug(entityClass.getSimpleName() + ".remove: " + entity);
			}
			// Before remove
			beforeDelete(entity);
			// Delete
			super.purge(entity);
			// After remove
			afterDelete(entity);
		}
	}


	@Override
	public void removeById(String id) {
		if (id != null) {
			if (log.isDebugEnabled()) {
				log.debug(entityClass.getSimpleName() + ".removeById: " + id);
			}
			// Get entity by id
			T entity = get(id);
			// Delete entity
			remove(entity);
		}
	}

	@Override
	public void unremoveById(String id) {
		if (id != null) {
			if (log.isDebugEnabled()) {
				log.debug(entityClass.getSimpleName() + ".unremoveById: " + id);
			}
			// Get entity by id
			T entity = get(id);
			// Delete entity
			unremove(entity);
		}
	}

	@Override
	public void purgeById(String id) throws PersistenceException {
		if (id != null) {
			if (log.isDebugEnabled()) {
				log.debug(entityClass.getSimpleName() + ".removeById: " + id);
			}
			// Get entity by id
			T entity = get(id);
			// Purge entity
			purge(entity);
		}
	}

	@Override
	public void remove(Collection<T> entities) throws PersistenceException {
		if (log.isDebugEnabled()) {
			log.debug("remove " + entities);
		}
		try {
			for (T entity : entities) {
				beforeDelete(entity);
				entity.setDeletionDate(System.currentTimeMillis());
			}
			db.executeBulk(entities);
			for (T entity : entities) {
				afterDelete(entity);
			}
		} catch (Exception e) {
			throw new PersistenceException("failed to remove entities ", e);
		}
	}

	@Override
	public void unremove(Collection<T> entities) throws PersistenceException {
		if (log.isDebugEnabled()) {
			log.debug("unremove " + entities);
		}
		try {
			for (T entity : entities) {
				beforeUnDelete(entity);
				entity.setDeletionDate(null);
			}
			db.executeBulk(entities);
			for (T entity : entities) {
				afterUnDelete(entity);
			}
		} catch (Exception e) {
			throw new PersistenceException("failed to unremove entities ", e);
		}
	}

	@Override
	public void purge(Collection<T> entities) throws PersistenceException {
		if (log.isDebugEnabled()) {
			log.debug("remove " + entities);
		}
		try {
			for (T entity : entities) {
				beforeDelete(entity);
			}
			db.executeBulk(entities.stream().map(BulkDeleteDocument::of).collect(Collectors.toList()));
			for (T entity : entities) {
				afterDelete(entity);
			}
		} catch (Exception e) {
			throw new PersistenceException("failed to remove entities ", e);
		}
	}

	@Override
	public void purgeByIds(Collection<String> ids) throws PersistenceException {
		purge(getList(ids));
	}

	@Override
	public void removeByIds(Collection<String> ids) throws PersistenceException {
		remove(getList(ids));
	}

	@Override
	public void unremoveByIds(Collection<String> ids) throws PersistenceException {
		unremove(getList(ids));
	}

	@Override
	public <K extends Collection<T>> K create(K entities) {
		return save(true, entities);
	}

	@Override
	public <K extends Collection<T>> K save(K entities) {
		return save(null, entities);
	}

	protected <K extends Collection<T>> K save(Boolean newEntity, K entities) {
		if (entities != null) {
			if (log.isDebugEnabled()) {
				log.debug(entityClass.getSimpleName() + ".save: " + entities.stream().filter(Objects::nonNull).map(entity->entity.getId() + ":" + entity.getRev()).collect(Collectors.joining(",")));
			}

			List<T> updatedEntities = new ArrayList<>();

			// Before save
			for (T entity : entities) {
				beforeSave(entity);

				// Check if key is missing and if this is a new entity
				boolean missingKey = (entity.getId() == null);
				newEntity = (newEntity != null) ? newEntity : (missingKey || entity.getRev()==null);

				// Add new key if missing
				if (missingKey) {
					keyManager.setNewKey(entity, entityClass.getSimpleName());
				}

				if (!newEntity) {
					updatedEntities.add(entity);
				}
			}

			//final Map<String,T> previousEntities = getList(updatedEntities.stream().map(T::getId).collect(Collectors.toList())).stream().collect(Collectors.toMap(T::getId,Function.<T>identity()));
			//updatedEntities.forEach((e)->saveRevHistory(e,previousEntities.get(e.getId())));

            // Save entity
			List<T> orderedEntities = new ArrayList<>(entities);
			List<DocumentOperationResult> results = db.executeBulk(orderedEntities);

			List<org.taktik.icure.exceptions.UpdateConflictException> conflicts = new ArrayList<>();
			for (DocumentOperationResult r : results) {
				if (r.getError() != null && r.getError().equals("conflict")) {
					conflicts.add(new org.taktik.icure.exceptions.UpdateConflictException(orderedEntities.stream().filter(e->e.getId().equals(r.getId())).findAny().orElse(null)));
				}
			}
			if (conflicts.size()>0) {
				throw new BulkUpdateConflictException(conflicts,orderedEntities);
			}

			orderedEntities.stream().filter(e -> e.getRev() != null).forEach(this::afterSave);
		}

		return entities;
	}

	@Override
	public void visitAll(final Function<T, Boolean> callback) throws PersistenceException {
		log.debug("visitAll");
		try {
			ViewQuery q = new ViewQuery()
					.allDocs()
					.includeDocs(true);
			db.queryView(q, entityClass).stream().forEach((t) -> callback.apply(refresh(t)));
		} catch (Exception e) {
			throw new PersistenceException("Failed to fetch all entites", e);
		}
	}

	protected void beforeDelete(T entity) {
	}

	protected void afterDelete(T entity) {
	}

	protected void beforeUnDelete(T entity) {
	}

	protected void afterUnDelete(T entity) {
	}

	public void postLoad(T entity) {
		doFetchRelationship(entity);
	}

	protected void doFetchRelationship(T object) {
	}

	@Override
	public void refreshIndex() {
		try {
			this.hasAny();
		} catch (Exception ignored) {}
	}

	@Override
	public List<String> getAllIds() {
		if (log.isDebugEnabled()) {
			log.debug(entityClass.getSimpleName() + ".getAllIds");
		}
		if (designDocContainsAllView()) {
			return db.queryView(createQuery("all").includeDocs(false), String.class);
		}

		return new ArrayList<>();
	}

    @Override
    public void warmupIndex(){
        if (log.isDebugEnabled()) {
            log.debug(entityClass.getSimpleName() + ".getAllIds");
        }
        if (designDocContainsAllView()) {
            db.queryView(createQuery("all").includeDocs(false).limit(1), String.class);
        }
    }

	private Options asEktorpOptions(Option... options) {
		Options ektorpOptions = new Options();
		for (Option option : options) {
			ektorpOptions.param(option.paramName(), "true");
		}
		return ektorpOptions;
	}

	public void initStandardDesignDocument(String groupId) {
		if (groupId==null || !(db instanceof CouchDbICureConnector)) {
			this.initStandardDesignDocument();
		} else {
			initDesignDocInternal(groupId,0);
		}
	}

	private void initDesignDocInternal(String groupId, int invocations) {
		CouchDbConnector cdb = (db instanceof CouchDbICureConnector) ? ((CouchDbICureConnector) db).getCouchDbICureConnector(groupId) : db;
		DesignDocument designDoc;
		if (cdb.contains(stdDesignDocumentId)) {
			designDoc = getDesignDocumentFactory().getFromDatabase(cdb, stdDesignDocumentId);
		} else {
			designDoc = getDesignDocumentFactory().newDesignDocumentInstance();
			designDoc.setId(stdDesignDocumentId);
		}
		log.debug("Generating DesignDocument for {}", type);
		DesignDocument generated = getDesignDocumentFactory().generateFrom(this);
		boolean changed = designDoc.mergeWith(generated, false);
		if (log.isDebugEnabled()) {
			debugDesignDoc(designDoc);
		}
		if (changed) {
			log.debug("DesignDocument changed or new. Updating database");
			try {
				cdb.update(designDoc);
			} catch (UpdateConflictException e) {
				log.warn("Update conflict occurred when trying to update design document: {}", designDoc.getId());
				if (invocations == 0) {
					backOff();
					log.info("retrying initStandardDesignDocument for design document: {}", designDoc.getId());
					initDesignDocInternal(groupId, 1);
				}
			}
		} else {
			log.debug("DesignDocument was unchanged. Database was not updated.");
		}
	}
}

File: src/main/kotlin/org/taktik/icure/asyncdao/impl/GenericDAOImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asyncdao.impl

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.*
import org.apache.commons.lang3.ArrayUtils
import org.slf4j.LoggerFactory
import org.taktik.couchdb.*
import org.taktik.couchdb.dao.designDocName
import org.taktik.couchdb.entity.DesignDocument
import org.taktik.couchdb.entity.Option
import org.taktik.couchdb.entity.View
import org.taktik.couchdb.exception.DocumentNotFoundException
import org.taktik.couchdb.exception.UpdateConflictException
import org.taktik.couchdb.id.IDGenerator
import org.taktik.couchdb.support.StdDesignDocumentFactory
import org.taktik.icure.asyncdao.GenericDAO
import org.taktik.icure.asyncdao.VersionedDesignDocumentQueries
import org.taktik.icure.entities.base.StoredDocument
import org.taktik.icure.exceptions.BulkUpdateConflictException
import org.taktik.icure.exceptions.PersistenceException
import org.taktik.icure.properties.CouchDbProperties
import java.net.URI
import java.nio.ByteBuffer

@FlowPreview
@ExperimentalCoroutinesApi
abstract class GenericDAOImpl<T : StoredDocument>(couchDbProperties: CouchDbProperties,
                                                  protected override val entityClass: Class<T>, protected val couchDbDispatcher: CouchDbDispatcher, protected val idGenerator: IDGenerator
) : GenericDAO<T>, VersionedDesignDocumentQueries<T>(entityClass, couchDbProperties) {
    private val log = LoggerFactory.getLogger(this.javaClass)
    protected val dbInstanceUrl = URI(couchDbProperties.url)

    override suspend fun contains(id: String): Boolean {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)

        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".contains: " + id)
        }
        return client.get(id, entityClass) != null
    }

    override suspend fun hasAny(): Boolean {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        return designDocContainsAllView(dbInstanceUrl) && client.queryView<String, String>(createQuery(client,"all").limit(1)).count() > 0
    }

    private suspend fun designDocContainsAllView(dbInstanceUrl: URI): Boolean {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        return client.get<DesignDocument>(designDocName(entityClass))?.views?.containsKey("all")
                ?: false
    }

    override fun getEntityIds(limit: Int?): Flow<String> = flow {
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".getAllIds")
        }
        if (designDocContainsAllView(dbInstanceUrl)) {
            val client = couchDbDispatcher.getClient(dbInstanceUrl)
            client.queryView<String, String>(if (limit != null) createQuery(client,"all", entityClass).limit(limit) else createQuery(client,"all", entityClass)).onEach { emit(it.id) }.collect()
        }
    }

    override fun getEntities(): Flow<T> = flow {
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".getAll")
        }
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        emitAll(
            client.queryView(createQuery(client, "all").includeDocs(true), String::class.java, String::class.java, entityClass).map { (it as? ViewRowWithDoc<*, *, T?>)?.doc }.filterNotNull()
        )
    }

    override fun getAttachment(documentId: String, attachmentId: String, rev: String?): Flow<ByteBuffer> = flow {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        emitAll(client.getAttachment(documentId, attachmentId, rev))
    }

    override suspend fun createAttachment(documentId: String, attachmentId: String, rev: String, contentType: String, data: Flow<ByteBuffer>): String {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        return client.createAttachment(documentId, attachmentId, rev, contentType, data)
    }

    override suspend fun deleteAttachment(documentId: String, rev: String, attachmentId: String): String {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        return client.deleteAttachment(documentId, attachmentId, rev)
    }

    override suspend fun get(id: String, vararg options: Option): T? = get(id, null, *options)

    override suspend fun get(id: String, rev: String?, vararg options: Option): T? {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".get: " + id + " [" + ArrayUtils.toString(options) + "]")
        }
        try {
            return rev?.let { client.get(id, it, entityClass, *options) }
                    ?: client.get(id, entityClass, *options)?.let { postLoad(it) }
        } catch (e: DocumentNotFoundException) {
            log.warn("Document not found", e)
        }

        return null
    }

    override fun getEntities(ids: Collection<String>): Flow<T> = flow {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".get: " + ids)
        }
        emitAll(client.get(ids, entityClass).map { this@GenericDAOImpl.postLoad(it) })
    }

    override fun getEntities(ids: Flow<String>): Flow<T> = flow  {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".get: " + ids)
        }
        emitAll(client.get(ids, entityClass).map { this@GenericDAOImpl.postLoad(it) })
    }

    override suspend fun create(entity: T): T? {
        return save(true, entity)
    }

    override suspend fun save(entity: T): T? {
        return save(null, entity)
    }

    protected open suspend fun save(newEntity: Boolean?, entity: T): T? {
        var newEntity = newEntity
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".save: " + entity.id + ":" + entity.rev)
        }

        return beforeSave(entity).let { entity ->
            if (newEntity ?: entity.rev == null) {
                client.create(entity, entityClass)
            } else {
                client.update(entity, entityClass)
                //saveRevHistory(entity, null);
            }.let { afterSave(it) }
        }
    }

    protected open suspend fun beforeSave(entity: T): T {
        return entity
    }

    protected open suspend fun afterSave(entity: T): T {
        return entity
    }

    protected open suspend fun beforeDelete(entity: T): T {
        return entity
    }

    protected open suspend fun afterDelete(entity: T): T {
        return entity
    }

    protected open suspend fun beforeUnDelete(entity: T): T {
        return entity
    }

    protected open suspend fun afterUnDelete(entity: T): T {
        return entity
    }

    protected open suspend fun postLoad(entity: T): T {
        return entity
    }

    override suspend fun remove(entity: T): DocIdentifier {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".remove: " + entity)
        }
        val deleted = client.update(beforeDelete(entity).withDeletionDate(deletionDate = System.currentTimeMillis()) as T, entityClass).let {
            afterDelete(it)
        }

        return DocIdentifier(deleted.id, deleted.rev)
    }

    override suspend fun unRemove(entity: T): DocIdentifier {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".remove: " + entity)
        }
        // Before remove
        return beforeUnDelete(entity).let {
            val undeleted = client.update(it.withDeletionDate(null) as T, entityClass).let {
                afterUnDelete(it)
            }

            DocIdentifier(undeleted.id, undeleted.rev)
        }
    }

    override suspend fun purge(entity: T): DocIdentifier {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".remove: " + entity)
        }
        // Delete
        val purged = client.delete(beforeDelete(entity))
        // After remove
        afterDelete(entity)
        return purged
    }

    override fun remove(entities: Collection<T>) = flow {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug("remove $entities")
        }
        try {
            val bulkUpdateResults = client.bulkUpdate(entities.map {
                beforeDelete(it).let {
                    it.withDeletionDate(System.currentTimeMillis()) as T
                }
            }, entityClass).onEach { r ->
                entities.firstOrNull { e -> r.id == e.id }?.let { afterDelete(it) }
            }
            emitAll(bulkUpdateResults.map { DocIdentifier(it.id, it.rev) })
        } catch (e: Exception) {
            throw PersistenceException("failed to remove entities ", e)
        }
    }

    override fun unRemove(entities: Collection<T>) = flow {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug("remove $entities")
        }
        try {
            val bulkUpdateResults = client.bulkUpdate(entities.map {
                beforeUnDelete(it).let {
                    it.withDeletionDate(null) as T
                }
            }, entityClass).onEach { r ->
                entities.firstOrNull { e -> r.id == e.id }?.let { afterUnDelete(it) }
            }
            emitAll(bulkUpdateResults.map { DocIdentifier(it.id, it.rev) })
        } catch (e: Exception) {
            throw PersistenceException("failed to remove entities ", e)
        }
    }

    // This function is not reactive, but it doesn't seem to be used at all anyway...
    override suspend fun purge(entities: Collection<T>) {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug("remove $entities")
        }
        try {
            val bulkDeleteResults = client.bulkDelete(entities.map { beforeDelete(it) }).toList()
            for (entity in entities) {
                afterDelete(entity)
            }
        } catch (e: Exception) {
            throw PersistenceException("failed to remove entities ", e)
        }
    }

    override fun <K : Collection<T>> create(entities: K): Flow<T> {
        return save(true, entities)
    }

    override fun <K : Collection<T>> save(entities: K): Flow<T> {
        return save(false, entities)
    }

    // TODO SH later: make sure this is correct
    protected open fun <K : Collection<T>> save(newEntity: Boolean?, entities: K): Flow<T> = flow {
        val client = couchDbDispatcher.getClient(dbInstanceUrl)
        if (log.isDebugEnabled) {
            log.debug(entityClass.simpleName + ".save: " + entities.mapNotNull { entity -> entity.id + ":" + entity.rev })
        }

        // Save entity
        val orderedEntities = ArrayList(entities.map { beforeSave(it) })
        val updatedEntities = orderedEntities.filter { !(newEntity ?: it.rev == null) }

        val results = client.bulkUpdate(orderedEntities, entityClass).toList()

        val conflicts = results.filter { it.error == "conflict" }.map { r -> UpdateConflictException(r.id, r.rev ?: "unknown") }.toList()
        if (conflicts.isNotEmpty()) {
            throw BulkUpdateConflictException(conflicts, orderedEntities)
        }
        emitAll(results.asFlow().mapNotNull { r ->
            (updatedEntities.firstOrNull { u -> r.id == u.id }
                    ?: entities.firstOrNull { u -> r.id == u.id })?.let { e -> r.rev?.let { e.withIdRev(rev = it) as T } ?: e }
        }.map { afterSave(it) })
    }

    override suspend fun forceInitStandardDesignDocument(updateIfExists: Boolean, useVersioning: Boolean) {
        forceInitStandardDesignDocument(couchDbDispatcher.getClient(dbInstanceUrl), updateIfExists, useVersioning)
    }

    override suspend fun forceInitStandardDesignDocument(client: Client, updateIfExists: Boolean, useVersioning: Boolean) {
        val designDocId = designDocName(this.entityClass)
        val generated = StdDesignDocumentFactory().generateFrom(designDocId, this)

        val fromDatabase = client.get(generated.id, DesignDocument::class.java) ?: client.get(designDocId, DesignDocument::class.java)

        val (merged, changed) = fromDatabase?.mergeWith(generated, true) ?: generated to true
        if (changed && (updateIfExists || fromDatabase == null)) {
            client.update(fromDatabase?.let { if (it.id == generated.id) merged.copy(rev = it.rev) else merged } ?: merged)
        }
    }

    override suspend fun initSystemDocumentIfAbsent(dbInstanceUrl: URI) {
        initSystemDocumentIfAbsent(couchDbDispatcher.getClient(dbInstanceUrl))
    }

    override suspend fun initSystemDocumentIfAbsent(client: Client) {
        val designDocId = designDocName("_System")
        val designDocument = client.get(designDocId, DesignDocument::class.java)
        if (designDocument == null) {
            client.update(DesignDocument(designDocId, views = mapOf("revs" to View("function (doc) { emit(doc.java_type, doc._rev); }"))))
        }
    }
}
